File: arraysAndHashing/products-of-array-except-self/csharp/ProductOfArrayExceptSelf.cs
using System;
using System.Collections.Generic;


public class ProductOfArrayExceptSelf
{
    public static void Main(string[] args)
    {
        ProductOfArrayExceptSelf solution = new ProductOfArrayExceptSelf();
        int[] nums = { 1, 2, 4, 6 };
        int[] result = solution.GetProductExceptSelf(nums);

        Console.WriteLine("Product of array except self:");
        foreach (int num in result)
        {
            Console.Write(num + " ");
        }
    }
    public int[] GetProductExceptSelf(int[] nums)
    {
        int n = nums.Length;
        int[] result = new int[n];
        int[] leftProducts = new int[n];
        int[] rightProducts = new int[n];

        leftProducts[0] = 1;
        for (int i = 1; i < n; i++)
        {
            leftProducts[i] = leftProducts[i - 1] * nums[i - 1];
        }

        rightProducts[n - 1] = 1;
        for (int i = n - 2; i >= 0; i--)
        {
            rightProducts[i] = rightProducts[i + 1] * nums[i + 1];
        }

        for (int i = 0; i < n; i++)
        {
            result[i] = leftProducts[i] * rightProducts[i];
        }

        return result;
    }
}
  

File: arraysAndHashing/products-of-array-except-self/python/ProductOfArrayExceptSelf.py
class ProductOfArrayExceptSelf:
    def __init__(self, nums):
        self.nums = nums

    def product_except_self(self):
        length = len(self.nums)
        answer = [1] * length

        left_product = 1
        for i in range(length):
            answer[i] = left_product
            left_product *= self.nums[i]

        right_product = 1
        for i in range(length - 1, -1, -1):
            answer[i] *= right_product
            right_product *= self.nums[i]

        return answer

if __name__ == "__main__":
    nums = [1, 2, 4, 6]
    solution = ProductOfArrayExceptSelf(nums)
    print(solution.product_except_self())